test/syncenginetestutils.h

```diff
@@ -0,0 +1,658 @@
+        setHeader(QNetworkRequest::ContentLengthHeader, payload.size());
+        setAttribute(QNetworkRequest::HttpStatusCodeAttribute, 200);
+        setRawHeader("OC-ETag", fileInfo->etag.toLatin1());
+        setRawHeader("ETag", fileInfo->etag.toLatin1());
+        emit metaDataChanged();
+        if (bytesAvailable())
+            emit readyRead();
+        emit finished();
+    }
+    void abort() override { }
+    qint64 bytesAvailable() const override { return payload.size() + QIODevice::bytesAvailable(); }
+
+    qint64 readData(char *data, qint64 maxlen) override {
+        qint64 len = std::min(qint64{payload.size()}, maxlen);
+        strncpy(data, payload.constData(), len);
+        payload.remove(0, len);
+        return len;
+    }
+
+};
+
+class FakeErrorReply : public QNetworkReply
+{
+    Q_OBJECT
+public:
+    FakeErrorReply(QNetworkAccessManager::Operation op, const QNetworkRequest &request, QObject *parent)
+    : QNetworkReply{parent} {
+        setRequest(request);
+        setUrl(request.url());
+        setOperation(op);
+        open(QIODevice::ReadOnly);
+        QMetaObject::invokeMethod(this, "respond", Qt::QueuedConnection);
+    }
+
+    Q_INVOKABLE void respond() {
+        setAttribute(QNetworkRequest::HttpStatusCodeAttribute, 500);
+        emit metaDataChanged();
+        emit finished();
+    }
+    void abort() override { }
+    qint64 readData(char *, qint64) override { return 0; }
+};
+
+class FakeQNAM : public QNetworkAccessManager
+{
+    FileInfo _remoteRootFileInfo;
+    QStringList _errorPaths;
+public:
+    FakeQNAM(FileInfo initialRoot) : _remoteRootFileInfo{std::move(initialRoot)} { }
+    FileInfo &currentRemoteState() { return _remoteRootFileInfo; }
+    QStringList &errorPaths() { return _errorPaths; }
+
+protected:
+    QNetworkReply *createRequest(Operation op, const QNetworkRequest &request,
+                                         QIODevice *outgoingData = 0) {
+        const QString fileName = request.url().path().mid(sRootUrl.path().length());
+        if (_errorPaths.contains(fileName))
+            return new FakeErrorReply{op, request, this};
+
+        auto verb = request.attribute(QNetworkRequest::CustomVerbAttribute);
+        if (verb == QLatin1String("PROPFIND"))
+            // Ignore outgoingData always returning somethign good enough, works for now.
+            return new FakePropfindReply{_remoteRootFileInfo, op, request, this};
+        else if (verb == QLatin1String("GET"))
+            return new FakeGetReply{_remoteRootFileInfo, op, request, this};
+        else if (verb == QLatin1String("PUT"))
+            return new FakePutReply{_remoteRootFileInfo, op, request, outgoingData->readAll(), this};
+        else if (verb == QLatin1String("MKCOL"))
+            return new FakeMkcolReply{_remoteRootFileInfo, op, request, this};
+        else if (verb == QLatin1String("DELETE"))
+            return new FakeDeleteReply{_remoteRootFileInfo, op, request, this};
```

Comment:
The "fake" comes from the "fake object" term used in testing. If I understand the term correctly myself, they aren't real replies, and that they do more than mocking, they implement a seam which you can use as a test harness.